src/cpu/testers/spatter_gen/spatter_gen.cc

```diff
@@ -0,0 +1,584 @@
+        owner->recvReqRetry();
+    }
+}
+
+void
+SpatterGen::recvReqRetry()
+{
+    if (nextSendEvent.pending()) {
+        nextSendEvent.wake();
+        scheduleNextSendEvent(nextCycle());
+    }
+}
+
+bool
+SpatterGen::SpatterGenPort::recvTimingResp(PacketPtr pkt) {
+    return owner->recvTimingResp(pkt);
+}
+
+bool
+SpatterGen::recvTimingResp(PacketPtr pkt)
+{
+    DPRINTF(SpatterGen, "%s: Received pkt: %s.\n", __func__, pkt->print());
+    assert(pkt->isResponse());
+
+    AccessInfo* acc_info = pkt->findNextSenderState<AccessInfo>();
+    if (acc_info->access_type() == AccessType::INDEX) {
+        assert(pkt->isRead());
+        // record trip time
+        Tick trip_time = (curTick() - requestDepartureTime[pkt->req]);
+        // trip time should be 0 if it's response for access to indexer.
+        assert(acc_info->trip_time() == 0);
+        acc_info->trip_time(trip_time);
+        // free memory
+        requestDepartureTime.erase(pkt->req);
+
+        stats.numIndexerReads++;
+        stats.indexerBytesRead += pkt->getSize();
+        stats.totalIndexerReadLatency += trip_time;
+
+        stats.indexerAccessLatency.sample(trip_time);
+        // TODO: record travel time for indexer accesses.
+        receiveBuffer.push(pkt);
+        receiveBufferInsertionTime.push(curTick());
+        // REMEMBER: might need to wake nextGenEvent up?
+        // IDEA: Separate pending to pendingInput and pendingOutput?
+        // I think for now we should not wake up nextGenEvent,
+        // because nextGenEvent does not sleep if there are not
+        // enough registers available, which is something that
+        // this scope of the if does not influence.
+    } else if (acc_info->access_type() == AccessType::VALUE) {
+        // trip time for values and total trip time (indexer + values)
+        Tick trip_time = curTick() - requestDepartureTime[pkt->req];
+        Tick tot_trip_time = trip_time + acc_info->trip_time();
+
+        requestDepartureTime.erase(pkt->req);
+        stats.valuesAccessLatency.sample(trip_time);
+        stats.totalIndirectAccessLatency.sample(tot_trip_time);
+        if (acc_info->kernel_type() == SpatterKernelType::gather) {
+            stats.numValuesReads++;
+            stats.valuesBytesRead += pkt->getSize();
+            stats.totalValuesReadLatency += trip_time;
+        } else if (acc_info->kernel_type() == SpatterKernelType::scatter) {
+            stats.numValuesWrites++;
+            stats.valuesBytesWritten += pkt->getSize();
+            stats.totalValuesWriteLatency += trip_time;
+        } else {
+            panic("Unknown access type.");
+        }
+        // CAUTION: We're going to decrement fpRegUsed here,
+        // it could cause inaccuracies if processNextGenEvent
+        // is called after recvTimingResp.
+        // it may be fine though since we don't simulate the rest,
+        // it's probably not going to ever be an issue since
+        // fpRegFileSize is probably >> requestBufferSize
```

Comment:
Would it make sense to put a panic/assert here?

Reply:
I will put a panic telling user to make fpRegFileSize bigger than request_buffer_entries.